File: src/systems/EconomyScenarioSimulator.js
import { EconomySystem } from './EconomySystem.js';
import { ECONOMY_BALANCE } from './EconomyBalance.js';

export const ECONOMY_SCENARIO_ACTIONS = Object.freeze({
  BUILD: 'BUILD',
  FINE: 'FINE',
  MISSION: 'MISSION',
  SPEND: 'SPEND'
});

export const BALANCED_SESSION_EVENTS = Object.freeze([
  Object.freeze({ minute: 5, type: 'BUILD', id: 'CYBERCAFE', cost: 200_000, incomePerMinute: 2_400 }),
  Object.freeze({ minute: 10, type: 'MISSION', id: 'session-contract-1', reward: 45_000 }),
  Object.freeze({ minute: 25, type: 'BUILD', id: 'METRO_LOFTS', cost: 350_000, incomePerMinute: 1_900 }),
  Object.freeze({ minute: 30, type: 'MISSION', id: 'session-contract-2', reward: 52_000 }),
  Object.freeze({ minute: 45, type: 'SPEND', id: 'incident-response-1', amount: 6_000, source: 'incident-response' }),
  Object.freeze({ minute: 50, type: 'MISSION', id: 'session-contract-3', reward: 68_000 }),
  Object.freeze({ minute: 70, type: 'BUILD', id: 'SOLAR_GRID', cost: 280_000, incomePerMinute: -450 }),
  Object.freeze({ minute: 75, type: 'MISSION', id: 'session-contract-4', reward: 75_000 }),
  Object.freeze({ minute: 90, type: 'MISSION', id: 'session-contract-5', reward: 85_000 }),
  Object.freeze({ minute: 105, type: 'FINE', id: 'session-fine-1', amount: 60_000 }),
  Object.freeze({ minute: 110, type: 'BUILD', id: 'CYBER_FAB', cost: 390_000, incomePerMinute: 6_200 })
]);

function freeze(value) {
  if (!value || typeof value !== 'object' || Object.isFrozen(value)) return value;
  Object.values(value).forEach(freeze);
  return Object.freeze(value);
}

function applyEvent(economy, event) {
  switch (event.type) {
    case ECONOMY_SCENARIO_ACTIONS.BUILD: {
      const spec = { id: event.id, incomePerMinute: event.incomePerMinute };
      if (!economy.spend(event.cost, { source: 'building-placement', referenceId: event.id, spec })) {
        throw new Error(`Scenario could not fund ${event.id} at minute ${event.minute}`);
      }
      economy.registerBuilding({
        id: `scenario:${event.id}`,
        grossIncomeRate: Math.max(0, event.incomePerMinute) / 60,
        operatingCostRate: Math.max(0, -event.incomePerMinute) / 60
      });
      break;
    }
    case ECONOMY_SCENARIO_ACTIONS.MISSION:
      economy.completeMission({ id: event.id, reward: event.reward, narrativeProgressDelta: 1 });
      break;
    case ECONOMY_SCENARIO_ACTIONS.FINE:
      economy.applyFine(event.amount, { referenceId: event.id });
      break;
    case ECONOMY_SCENARIO_ACTIONS.SPEND:
      if (!economy.spend(event.amount, { source: event.source, referenceId: event.id })) {
        throw new Error(`Scenario debit ${event.id} was rejected at minute ${event.minute}`);
      }
      break;
    default:
      throw new RangeError(`Unsupported economy scenario action: ${String(event.type)}`);
  }
}

/**
 * Runs ordered player/economy actions against the production EconomySystem.
 * Explicit simulation time and stable ordering make the same scenario exactly
 * reproducible in tests, CI, and future balance tooling.
 */
export function runEconomyScenario({
  durationMinutes,
  events = [],
  initialTreasury = ECONOMY_BALANCE.startingTreasury,
  passiveIncomeRate = ECONOMY_BALANCE.baseRevenuePerSecond
} = {}) {
  if (!Number.isFinite(durationMinutes) || durationMinutes < 0) {
    throw new RangeError('durationMinutes must be a non-negative finite number');
  }
  const normalizedEvents = events.map((event, index) => {
    if (!event || typeof event !== 'object' || !Number.isFinite(event.minute) || event.minute < 0) {
      throw new TypeError(`scenario event ${index} requires a non-negative finite minute`);
    }
    return { ...event, index };
  });
  const ordered = normalizedEvents
    .filter(event => event.minute <= durationMinutes)
    .sort((left, right) => left.minute - right.minute || left.index - right.index);
  const economy = new EconomySystem({ initialTreasury, passiveIncomeRate });
  const transactions = [];
  let elapsedMinutes = 0;

  for (const event of ordered) {
    economy.update((event.minute - elapsedMinutes) * 60);
    applyEvent(economy, event);
    elapsedMinutes = event.minute;
    transactions.push({
      minute: event.minute,
      type: event.type,
      id: event.id,
      treasury: economy.treasury,
      fiscalStatus: economy.snapshot().fiscalStatus
    });
  }
  economy.update((durationMinutes - elapsedMinutes) * 60);
  const snapshot = economy.snapshot();
  return freeze({
    durationMinutes,
    snapshot,
    assetCount: snapshot.buildings.length,
    missionCount: snapshot.completedMissions.length,
    transactions
  });
}

export function runBalancedSession(durationMinutes) {
  return runEconomyScenario({ durationMinutes, events: BALANCED_SESSION_EVENTS });
}

export default runEconomyScenario;

File: godot/MetroPulse.Domain/Economy/EconomyScenarioSimulator.cs
using MetroPulse.Domain.Content;

namespace MetroPulse.Domain.Economy;

public static class EconomyScenarioActions
{
    public const string Build = "BUILD";
    public const string Fine = "FINE";
    public const string Mission = "MISSION";
    public const string Spend = "SPEND";
}

public sealed record EconomyScenarioEvent(
    double Minute,
    string Type,
    string Id,
    double Amount = 0,
    double IncomePerMinute = 0,
    string Source = "manual");

public sealed record EconomyScenarioTransaction(
    double Minute,
    string Type,
    string Id,
    double Treasury,
    string FiscalStatus);

public sealed record EconomyScenarioResult(
    double DurationMinutes,
    EconomyLedgerSnapshot Snapshot,
    int AssetCount,
    int MissionCount,
    IReadOnlyList<EconomyScenarioTransaction> Transactions);

public static class EconomyScenarioSimulator
{
    public static readonly IReadOnlyList<EconomyScenarioEvent> BalancedSessionEvents = Array.AsReadOnly(new EconomyScenarioEvent[]
    {
        new(5, EconomyScenarioActions.Build, "CYBERCAFE", Amount: 200_000, IncomePerMinute: 2_400),
        new(10, EconomyScenarioActions.Mission, "session-contract-1", Amount: 45_000),
        new(25, EconomyScenarioActions.Build, "METRO_LOFTS", Amount: 350_000, IncomePerMinute: 1_900),
        new(30, EconomyScenarioActions.Mission, "session-contract-2", Amount: 52_000),
        new(45, EconomyScenarioActions.Spend, "incident-response-1", Amount: 6_000, Source: "incident-response"),
        new(50, EconomyScenarioActions.Mission, "session-contract-3", Amount: 68_000),
        new(70, EconomyScenarioActions.Build, "SOLAR_GRID", Amount: 280_000, IncomePerMinute: -450),
        new(75, EconomyScenarioActions.Mission, "session-contract-4", Amount: 75_000),
        new(90, EconomyScenarioActions.Mission, "session-contract-5", Amount: 85_000),
        new(105, EconomyScenarioActions.Fine, "session-fine-1", Amount: 60_000),
        new(110, EconomyScenarioActions.Build, "CYBER_FAB", Amount: 390_000, IncomePerMinute: 6_200),
    });

    public static EconomyScenarioResult RunBalancedSession(double durationMinutes)
    {
        EconomyBalanceDefinition balance = GameContentRegistry.LoadProduction().EconomyBalance;
        return Run(
            durationMinutes,
            BalancedSessionEvents,
            balance,
            balance.StartingTreasury,
            balance.BaseRevenuePerSecond);
    }

    public static EconomyScenarioResult Run(
        double durationMinutes,
        IEnumerable<EconomyScenarioEvent> events,
        EconomyBalanceDefinition balance,
        double initialTreasury,
        double passiveIncomeRate)
    {
        if (!double.IsFinite(durationMinutes) || durationMinutes < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(durationMinutes));
        }
        ArgumentNullException.ThrowIfNull(events);
        ArgumentNullException.ThrowIfNull(balance);

        EconomyScenarioEvent[] materializedEvents = events.ToArray();
        for (int index = 0; index < materializedEvents.Length; index += 1)
        {
            EconomyScenarioEvent scenarioEvent = materializedEvents[index]
                ?? throw new ArgumentException($"Scenario event {index} is required.", nameof(events));
            if (!double.IsFinite(scenarioEvent.Minute) || scenarioEvent.Minute < 0)
            {
                throw new ArgumentException(
                    $"Scenario event {index} requires a non-negative finite minute.",
                    nameof(events));
            }
        }

        var ledger = new EconomyLedger(balance, initialTreasury, passiveIncomeRate);
        var transactions = new List<EconomyScenarioTransaction>();
        double elapsedMinutes = 0;
        IEnumerable<EconomyScenarioEvent> ordered = materializedEvents
            .Select((scenarioEvent, index) => (Event: scenarioEvent, Index: index))
            .Where(item => item.Event.Minute <= durationMinutes)
            .OrderBy(item => item.Event.Minute)
            .ThenBy(item => item.Index)
            .Select(item => item.Event);

        foreach (EconomyScenarioEvent scenarioEvent in ordered)
        {
            ledger.Update((scenarioEvent.Minute - elapsedMinutes) * 60);
            ApplyEvent(ledger, scenarioEvent);
            elapsedMinutes = scenarioEvent.Minute;
            transactions.Add(new EconomyScenarioTransaction(
                scenarioEvent.Minute,
                scenarioEvent.Type,
                scenarioEvent.Id,
                ledger.Treasury,
                ledger.GetFiscalOverview().Status));
        }

        ledger.Update((durationMinutes - elapsedMinutes) * 60);
        EconomyLedgerSnapshot snapshot = ledger.Snapshot();
        return new EconomyScenarioResult(
            durationMinutes,
            snapshot,
            snapshot.Buildings.Count,
            snapshot.CompletedMissions.Count,
            Array.AsReadOnly(transactions.ToArray()));
    }

    private static void ApplyEvent(EconomyLedger ledger, EconomyScenarioEvent scenarioEvent)
    {
        switch (scenarioEvent.Type)
        {
            case EconomyScenarioActions.Build:
                {
                    var spec = new EconomyInvestmentSpec { IncomePerMinute = scenarioEvent.IncomePerMinute };
                    bool funded = ledger.Spend(scenarioEvent.Amount, new SpendingContext
                    {
                        Source = "building-placement",
                        ReferenceId = scenarioEvent.Id,
                        Spec = spec,
                    });
                    if (!funded)
                    {
                        throw new InvalidOperationException(
                            $"Scenario could not fund {scenarioEvent.Id} at minute {scenarioEvent.Minute}.");
                    }
                    ledger.RegisterBuilding(new EconomyBuilding(
                        $"scenario:{scenarioEvent.Id}",
                        Math.Max(0, scenarioEvent.IncomePerMinute) / 60,
                        Math.Max(0, -scenarioEvent.IncomePerMinute) / 60));
                    return;
                }
            case EconomyScenarioActions.Mission:
                ledger.CompleteMission(scenarioEvent.Id, scenarioEvent.Amount);
                return;
            case EconomyScenarioActions.Fine:
                ledger.ApplyFine(scenarioEvent.Amount, referenceId: scenarioEvent.Id);
                return;
            case EconomyScenarioActions.Spend:
                if (!ledger.Spend(scenarioEvent.Amount, new SpendingContext
                {
                    Source = scenarioEvent.Source,
                    ReferenceId = scenarioEvent.Id,
                }))
                {
                    throw new InvalidOperationException(
                        $"Scenario debit {scenarioEvent.Id} was rejected at minute {scenarioEvent.Minute}.");
                }
                return;
            default:
                throw new ArgumentOutOfRangeException(
                    nameof(scenarioEvent),
                    scenarioEvent.Type,
                    "Unsupported economy scenario action.");
        }
    }
}
